### ISSUE TITLE
cmd/go: TestScript/version failing on Plan 9

### ISSUE THREAD

**Issue by 0intro** [golang/go#31706]
CL [173343](https://golang.org/cl/173343) implemented `go version <binary>`. However, the [openExe](https://github.com/golang/go/blob/2177bfb343c4950be88536044bb38c90f05ad3ed/src/cmd/go/internal/version/exe.go#L30) function doesn't implement decoding of Plan 9 executables. Consequently, the TestScript/version test fails.

```
--- FAIL: TestScript (0.01s)
    --- FAIL: TestScript/version (1.84s)
        script_test.go:190: 
            > env GO111MODULE=on
            > go build -o fortune.exe rsc.io/fortune
            [stderr]
            go: finding rsc.io/fortune v1.0.0
            go: downloading rsc.io/fortune v1.0.0
            go: extracting rsc.io/fortune v1.0.0
            go: finding rsc.io/quote v1.5.2
            go: downloading rsc.io/quote v1.5.2
            go: extracting rsc.io/quote v1.5.2
            go: finding rsc.io/sampler v1.3.0
            go: finding golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c
            go: downloading rsc.io/sampler v1.3.0
            go: extracting rsc.io/sampler v1.3.0
            go: downloading golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c
            go: extracting golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c
            > go version fortune.exe
            [stderr]
            fortune.exe: unrecognized executable format
            > stdout '^fortune.exe: .+'
            FAIL: testdata/script/version.txt:4: no match for `(?m)^fortune.exe: .+` found in stdout
```

**Comment by vitaminniy** [comment]
Probably we can determine that file is plan9 executable with:
```go
bytes.HasPrefix(data, plan9exeHeader(data[:5]))

func plan9exeHeader(d []byte) []byte {
	result := make([]byte, 4)
	binary.BigEndian.PutUint32(result, binary.BigEndian.Uint32(d))
	return result
}
```

**Comment by 0intro** [comment]
I've started to implement parsing of Go version in Plan 9 a.out executables.

I can match the info magic `\xff Go buildinf:` and parse endianess and ptrSize successfully.
However, I fail at reading values of buildVersion and modinfo.

```
% go test -v -run 'TestScript/^version$'
=== RUN   TestScript
=== RUN   TestScript/version
=== PAUSE TestScript/version
=== CONT  TestScript/version
go test proxy starting
go proxy_test: invalid module path encoding "example.com/invalidpath/v1": malformed module path "example.com/invalidpath/v1": invalid version
go test proxy running at GOPROXY=http://127.0.0.1:43806/mod
--- FAIL: TestScript (0.00s)
    --- FAIL: TestScript/version (2.15s)
        script_test.go:190: 
            WORK=$WORK
            PATH=/tmp/cmd-go-test-102488855/tmpdir217311114/testbin
            home=/no-home
            CCACHE_DISABLE=1
            GOARCH=386
            GOCACHE=/usr/djc/lib/cache/go-build
            GOOS=plan9
            GOPATH=$WORK/gopath
            GOPROXY=http://127.0.0.1:43806/mod
            GOROOT=/usr/go
            GONOVERIFY=*
            TMPDIR=$WORK/tmp
            devnull=/dev/null
            goversion=1.13
            :=
            path=/tmp/cmd-go-test-102488855/tmpdir217311114/testbin/bin.
            exe=
            CC=8c
            GCCGO=gccgo
            
            > env GO111MODULE=on
            > go build -o fortune.exe rsc.io/fortune
            [stderr]
            go: finding rsc.io/fortune v1.0.0
            go: downloading rsc.io/fortune v1.0.0
            go: extracting rsc.io/fortune v1.0.0
            go: finding rsc.io/quote v1.5.2
            go: downloading rsc.io/quote v1.5.2
            go: extracting rsc.io/quote v1.5.2
            go: finding rsc.io/sampler v1.3.0
            go: finding golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c
            go: downloading rsc.io/sampler v1.3.0
            go: extracting rsc.io/sampler v1.3.0
            go: downloading golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c
            go: extracting golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c
            > go version fortune.exe
            [stdout]
            Header Bss 0x00014060
            Header Entry 0x00043820
            Header PtrSize 4
            Header LoadAddress 0x00001000
            Header HdrSize 32
            Section text 0x00000020 743971
            Section data 0x000b5a43 58560
            Section syms 0x000c3f03 64855
            Section spsz 0x000d3c5a 0
            Section pcsz 0x000d3c5a 0
            ReadData 0x000b5a43 65536
            match Go buildinf
            LittleEndian
            ptrSize 4
            ReadData 0x000c2b68 8
            ReadData 0x00000000 0
            vers ''
            ReadData 0x000c2b88 8
            ReadData 0x00000000 0
            mod ''
            [stderr]
            fortune.exe: go version not found
            > stdout '^fortune.exe: .+'
            FAIL: testdata/script/version.txt:4: no match for `(?m)^fortune.exe: .+` found in stdout
            
FAIL
exit status: 'go.test 547205: 1'
FAIL	cmd/go	8.536s
```

The buildVersion and modinfo addresses I get are identical to the values returned by the `nm` tool.

```
% nm fortune.exe | grep '(buildVersion|modinfo)'
   c2b68 D runtime.buildVersion
   c2b88 D runtime.modinfo
```

Either the buildVersion and modinfo values aren't set correctly or I did a mistake in the addresses calculations when reading the values.

@rsc In Plan 9 executables, data segment addresses are absolute to the beginning of the file, right?

**Comment by rsc** [comment]
> @rsc In Plan 9 executables, data segment addresses are absolute to the beginning of the file, right?

No. See http://man.cat-v.org/plan_9/6/a.out. There's a zeroed page at the start of the addr space to catch nil writes, and then some extra page alignment between text and data.

**Comment by oridb** [comment]
I believe this is obsolete; I think I fixed this in 0816d3871376f8fe029db1a6d00b1f6882106043
